XenAdmin/Dialogs/AddVGPUDialog.cs

```diff
@@ -68,29 +68,26 @@ private void PopulateComboBox()
                     var enabledTypes = _vm.Connection.ResolveAll(gpu_group.enabled_VGPU_types);
                     var allTypes = _vm.Connection.ResolveAll(gpu_group.supported_VGPU_types);
                     var disabledTypes = allTypes.FindAll(t => !enabledTypes.Exists(e => e.opaque_ref == t.opaque_ref));
-
-                    allTypes.Sort();
-                    allTypes.Reverse();
-
-                    var commonTypes = new List<VGPU_type>();
-                    commonTypes.AddRange(allTypes);
+                    HashSet<VGPU_type> commonTypesSet = new HashSet<VGPU_type>(allTypes);
 
                     foreach (var eVgpu in existingVGpus)
                     {
                         var etype = _vm.Connection.Resolve(eVgpu.type);
-                        foreach (var vgpuType in allTypes)
-                        {
-                            if (!etype.compatible_types_in_vm.Contains(vgpuType.model_name))
-                                commonTypes.Remove(vgpuType);
-                        }
+                        List<XenRef<VGPU_type>> existing_compatible_types_in_vm_refs = etype.compatible_types_in_vm;

```

Comment:
We could simplify the code more here, because we don't need to resolve `gpu_group.enabled_VGPU_types` and `gpu_group.supported_VGPU_types` in order to build the `commonTypes` list.
```
                    var commonTypesSet = new HashSet<XenRef<VGPU_type>>(gpu_group.supported_VGPU_types);

                    foreach (var eVgpu in existingVGpus)
                    {
                        var eType = _vm.Connection.Resolve(eVgpu.type);
                        if (eType != null && eType.compatible_types_in_vm != null)
                        {
                            var existingCompatibleTypesSet = new HashSet<XenRef<VGPU_type>>(eType.compatible_types_in_vm);
                            commonTypesSet.IntersectWith(existingCompatibleTypesSet);
                        }
                    }

                    var commonTypes = _vm.Connection.ResolveAll(commonTypesSet);
                    commonTypes.Sort();
                    commonTypes.Reverse();

                    if (gpu_group.HasVGpu() && commonTypes.Count > 0)
                        comboBoxTypes.Items.Add(new GpuTuple(gpu_group, _vm.Connection.ResolveAll(gpu_group.supported_VGPU_types).ToArray())); // Group item

                    var disabledTypes = _vm.Connection.ResolveAll(gpu_group.supported_VGPU_types.FindAll(t => !gpu_group.enabled_VGPU_types.Exists(e => e.opaque_ref == t.opaque_ref)));

                    foreach (var vgpuType in commonTypes)
                        comboBoxTypes.Items.Add(new GpuTuple(gpu_group, vgpuType, disabledTypes.ToArray())); // GPU_type item
```

Reply:
Great! 
For disabledTypes, maybe we can simplify with:
var disabledTypes = _vm.Connection.ResolveAll(gpu_group.supported_VGPU_types.Except(gpu_group.enabled_VGPU_types));
Will try it soon.

Reply:
New code for disabledTypes works.